### PR TITLE
Upgrade to latest OpenTelemetry and make other fixes

### PR DESCRIPTION
**Pull request by seancribbs**
* Uses `1.0.0-rc` versions of OpenTelemetry
* Restructures span creation / termination to be dependent on the `Plug.Router` telemetry events, instead of the optional `Plug.Telemetry` events.
* Adds an optional Plug that emits propagation headers in the response.
* Updates span attributes to be more closely aligned with those in `opentelemetry_phoenix`.
* Restructures tests to cover more of the code and be less flakey.

### CODE REVIEW

File: lib/opentelemetry_plug.ex [hunk 1]
```diff
@@ -82,47 +78,64 @@ defmodule OpentelemetryPlug do
...
+      Span.record_exception(
+        Tracer.current_span_ctx(),
+        exception,
+        stacktrace
+      )
+
+      Tracer.set_status(OpenTelemetry.status(:error, Exception.message(exception)))
+      Tracer.set_attribute("http.status_code", 500)
+      Tracer.end_span()
+      restore_parent_ctx()
```
Comment (bryannaegele) [inline lib/opentelemetry_plug.ex:134]: You shouldn't need to set this. The plug will continue to run through the stop event where the end error code is already set. Also, the stop event would be where the span is getting closed, so you I don't think you should be ending the span or restoring the parent.

The flow (at least in Phoenix) is always one of these two:

* `start -> stop`
* `start -> exception -> stop`

File: mix.exs [hunk 2]
```diff
@@ -22,9 +22,9 @@ defmodule OpentelemetryPlug.MixProject do
   defp deps do
     [
       {:hackney, "~> 1.0", only: :test, runtime: false},
-      {:opentelemetry_api, "~> 0.3"},
-      {:opentelemetry, "~> 0.3"},
-      {:plug, "~> 1.10.1"},
+      {:opentelemetry_api, "~> 1.0.0-rc"},
+      {:opentelemetry, "~> 1.0.0-rc", only: :test},
+      {:plug, "~> 1.11.1"},
```
Comment (bryannaegele) [inline mix.exs:27]: We should attempt to keep this as relaxed as we possibly can. Did something change in 1.11 that this implementation relies on?

File: lib/opentelemetry_plug.ex [hunk 3]
```diff
@@ -82,47 +78,64 @@ defmodule OpentelemetryPlug do
...
       {"http.user_agent", user_agent},
       {"http.method", conn.method},
       {"net.peer.ip", to_string(:inet_parse.ntoa(peer_ip))},
       {"net.peer.port", peer_data.port},
       {"net.peer.name", host},
       {"net.transport", "IP.TCP"},
       {"net.host.ip", to_string(:inet_parse.ntoa(conn.remote_ip))},
-      {"net.host.port", conn.port} | optional_attributes(conn)
-      # {"net.host.name", HostName}
+      {"net.host.port", conn.port}
```
[the remark this answers is not in the document]
Comment (seancribbs) [reply lib/opentelemetry_plug.ex:90]: How important is it that they be atoms, other than syntactic noise of the current form?

File: lib/opentelemetry_plug.ex (hunk 1 above)
Comment (seancribbs) [reply lib/opentelemetry_plug.ex:134]: It appears to my eyes that on [`Plug.Router` it's always](https://github.com/elixir-plug/plug/blob/master/lib/plug/router.ex#L283-L303):
* `start -> stop`
* `start -> exception`

The `else` clause is never called after `catch`, so the `stop` event won't be invoked in that path.

I'll remove the context-restoring stuff.

File: lib/opentelemetry_plug.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (bryannaegele) [reply lib/opentelemetry_plug.ex:90]: Binaries will result in a ~40-60% increase in memory pressure in systems under load for a web server as observed in testing. These spans all go into and out of ETS tables, so they have to be copied, thus greatly increasing memory pressure. Whereas atom keys are referenced out of the atom table, thus only taking 1 word.

It is favorable to use atoms wherever possible as tracing can easily double the memory load of an app out of the box.

File: mix.exs (hunk 2 above)
Comment (seancribbs) [reply mix.exs:27]: The [plug router telemetry events were added after 1.10.1](https://github.com/elixir-plug/plug/commit/b753a2ffc24cdd41bab0cb576a5827954f8f5d0a#diff-8abb6a1c8f0384efb997eebfa1f2a8b3f91e339b95ba79de9a50132b3ddddd3c), thanks to @tsloughter.

File: lib/opentelemetry_plug.ex (hunk 3 above)
[the remark this answers is not in the document]
Comment (seancribbs) [reply lib/opentelemetry_plug.ex]: Resolved in 92b6304

File: lib/opentelemetry_plug.ex [hunk 4]
```diff
@@ -82,47 +78,64 @@ defmodule OpentelemetryPlug do
...
-    end
+    Tracer.set_current_span(span_ctx)
   end
 
   @doc false
   def handle_stop(_, _measurements, %{conn: conn}, _config) do
     if in_span?() do
-      OpenTelemetry.Span.set_attribute("http.status", conn.status)
-      OpenTelemetry.Tracer.end_span()
+      Tracer.set_attribute("http.status_code", conn.status)
```
[the remark this answers is not in the document]
Comment (seancribbs) [reply lib/opentelemetry_plug.ex]: Resolved in 92b6304

File: lib/opentelemetry_plug.ex (hunk 1 above)
Comment (seancribbs) [reply lib/opentelemetry_plug.ex]: Removed the context save/restore in 92b6304

File: lib/opentelemetry_plug.ex [hunk 5]
```diff
@@ -3,74 +3,70 @@ defmodule OpentelemetryPlug do
...
-    # TODO: add config for what paths are traced
-
+  def handle_start(_, _measurements, %{conn: conn, route: route}, _config) do
+    save_parent_ctx()
     # setup OpenTelemetry context based on request headers
-    :ot_propagation.http_extract(conn.req_headers)
+    :otel_propagator.text_map_extract(conn.req_headers)
 
-    span_name = "HTTP " <> conn.method
+    span_name = "#{conn.method} #{route}"
```
[the remark this answers is not in the document]
Comment (seancribbs) [reply lib/opentelemetry_plug.ex]: Resolved in 30c73dd

File: lib/opentelemetry_plug.ex [hunk 6]
```diff
@@ -100,42 +99,34 @@ defmodule OpentelemetryPlug do
...
+    Span.record_exception(
+      Tracer.current_span_ctx(),
+      exception,
+      stacktrace
+    )
 
-  defp in_span?, do: Tracer.current_span_ctx() != :undefined
+    Tracer.set_status(OpenTelemetry.status(:error, Exception.message(exception)))
+    Tracer.set_attribute(:"http.status_code", 500)
+    Tracer.end_span()
```
Comment (bryannaegele) [inline lib/opentelemetry_plug.ex:128]: The span will be ended and the status code set in the stop handler. These are safe to delete. :)

File: lib/opentelemetry_plug.ex [hunk 7]
```diff
@@ -100,42 +99,34 @@ defmodule OpentelemetryPlug do
...
-      # code the client failed to interpret, status MUST be set to Error.
-      #
-      # Don't set the span status description if the reason can be inferred from
-      # http.status_code.
-      if conn.status >= 400 do
-        Tracer.set_status(OpenTelemetry.status(:error, ""))
-      end
-
-      Tracer.end_span()
-      restore_parent_ctx()
```
Comment (bryannaegele) [inline lib/opentelemetry_plug.ex]: You do still need to manage the saving and restoring of the span ctx. That isn't done automatically anymore, so you had it right the first time. The `in_span?` deal is a relic from pre-ctx days

Comment (seancribbs) [reply lib/opentelemetry_plug.ex]: Woops.

File: lib/opentelemetry_plug.ex (hunk 6 above)
Comment (seancribbs) [reply lib/opentelemetry_plug.ex:128]: The `stop` handler is _not_ called in the exception case. See https://hexdocs.pm/elixir/Kernel.SpecialForms.html#try/1-else-clauses

> Only the result of the tried expression falls down to the `else` clause. If the `try` ends up in the `rescue` or `catch` clauses, their result will not fall down to `else`

File: mix.exs (hunk 2 above)
Comment (bryannaegele) [reply mix.exs:27]: Yeah, that commit was released in 1.10.1

https://github.com/elixir-plug/plug/compare/v1.10.0...v1.10.1

File: lib/opentelemetry_plug.ex (hunk 6 above)
Comment (bryannaegele) [reply lib/opentelemetry_plug.ex:130]: Ok. This is being handled differently than in Phoenix then. Phoenix will catch the error again and complete the plug pipeline back out.

File: mix.exs (hunk 2 above)
Comment (seancribbs) [reply mix.exs]: Relaxed in 89d05d2

**Review by bryannaegele — COMMENTED** [review]
I think we're close. A couple of items:

1. The lock file still has the updated Plug version.
2. Let's get some documentation in on the additional optional response propagation plug you're introducing, including a link to the warning in the W3C docs of using it. https://www.w3.org/TR/trace-context/#other-risks

**Comment by seancribbs** [comment]
> The lock file still has the updated Plug version.

Do you have a solution for this? If I use the constraint as-is and delete/unlock it, `mix deps.get` still fetches/locks 1.11.1.

> Let's get some documentation in on the additional optional response propagation plug you're introducing, including a link to the warning in the W3C docs of using it. https://www.w3.org/TR/trace-context/#other-risks

Thanks, I'll do that Monday.

**Comment by bryannaegele** [comment]
> Do you have a solution for this? If I use the constraint as-is and delete/unlock it, mix deps.get still fetches/locks 1.11.1.

Easiest way I know is to checkout the file again to reset it, then do a `deps.get` with just the otel deps updates in `mix.exs`.

**Comment by seancribbs** [comment]
> Easiest way I know is to checkout the file again to reset it, then do a `deps.get` with just the otel deps updates in `mix.exs`.
> 
> ```
> git checkout ^HEAD mix.lock
> mix deps.get
> ```

I ended up pinning the dependency locally to `1.10.1`, running `mix deps.get` to update the lock file, and then relaxing the constraint. It even found a bug (they had renamed a key in the `metadata` on exception).

**Review by bryannaegele — APPROVED** [review]

**Comment by seancribbs** [comment]
❤️

**Pull request merged**